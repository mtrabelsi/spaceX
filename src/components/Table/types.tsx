export type Link = {
    reddit: string | null,
    article: string | null,
    wikipedia: string | null
}

export type DataHistoryType = {
    id: string,
    title: string,
    event_date_utc: string,
    event_date_unix: number,
    details: string,
    links: Link
}

export type DataLaunchType = {
    flight_number: number,
    mission_name: string,
    launch_date_utc: string,
    launch_site: {
        site_name: string
    }
}

export type TableItemProps = {
    renderData: () => any,
    dataType: 'history' | 'launches',
    itemData: DataHistoryType | DataLaunchType,
    itemClickHandler: (id: string) => void
}
