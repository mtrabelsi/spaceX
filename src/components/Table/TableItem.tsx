import React from 'react';
import { TableItemProps } from './types';
import {
  StyledTableItem,
} from './atoms';

function TableItem(props : TableItemProps) {
  const { itemData, renderData, itemClickHandler } = props;
  return (
    <StyledTableItem
      onClick={(e: React.MouseEvent) => itemClickHandler()}
    >
      {renderData()}
    </StyledTableItem>
  );
}

export default TableItem;
