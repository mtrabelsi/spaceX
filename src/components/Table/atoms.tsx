import s from 'styled-components';
import { defaultColor, primarytColor, lightColor, secondaryColor, VerticalEllipsis } from '../Global'

export const StyledTableItem = s.section`
    border-bottom: 1px solid silver;
    padding: 2px;
`
export const Title = s.div`
    color: ${defaultColor};
    font-weight: bold;
`
export const EventDateUTC = s.div`
    color: ${lightColor};
`
export const EventDateUnix = s.div`
    color: ${defaultColor};
`
export const FlightNumber = s.div`
    color: ${secondaryColor};
`
export const Details = s.div`
    color: ${secondaryColor};
    ${VerticalEllipsis(2)};
`
export const Links = s.div`
    display: flex;
    justify-content: space-between;

    width: 100%;

`
export const Link = s.a`
    width: 100%;
    text-decoration: none;
    color: ${secondaryColor};
    font-weight: bold;
`
export const DataWrapper = s.div`
    padding: 5px;
    display: flex;
    
    justify-content: space-between;
`
export const OtherDataWrapper = s.div`
    display: flex;
    justify-content: space-between;
`