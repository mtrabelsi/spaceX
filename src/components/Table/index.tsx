import React from 'react'
import PropTypes from 'prop-types';
import { DataHistoryType, DataLaunchType } from './types'
import TableItem from './TableItem'
import { renderHistory, renderLaunches } from './index.helper';

type TableProps = {
    dataType: 'history' | 'launches',
    data: Array<DataHistoryType | DataLaunchType>,
    itemClickHandler: (id: string) => void
}

function Table(props : TableProps) {
    const { dataType, data, itemClickHandler } = props
    const dataLength : number =  data && data.length
    const historyArr = data as DataHistoryType[] 
    const launchesArr = data as DataLaunchType[]
    return (<section>
        {(dataType === 'history') && dataLength > 0 && historyArr.map((d: DataHistoryType) => (<TableItem
            key={d.id}
            renderData={() => renderHistory(d)}
            dataType={dataType}
            itemData={d} 
            itemClickHandler={itemClickHandler} 
        />))}
        {(dataType === 'launches') && dataLength > 0 && launchesArr.map((d: DataLaunchType) => (<TableItem
            key={d.flight_number}
            renderData={() => renderLaunches(d)}
            dataType={dataType}
            itemData={d} 
            itemClickHandler={itemClickHandler} 
        />))}
        {dataLength === 0 && 
        <div>No Item found!</div>}
    </section>)
}

Table.propTypes = {
    data: PropTypes.array
};

Table.defaultProps = {
    data: []
};
  
export default Table