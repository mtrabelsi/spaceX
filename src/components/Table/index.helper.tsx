import React from 'react'
import { DataWrapper, Title, EventDateUnix, EventDateUTC, Details, Links, Link } from './atoms'
import { DataHistoryType, DataLaunchType } from './types'

const LinkWithTaget = (props: any) => <Link {...props} target="_blank">{props.children}</Link>
const getAlign = (arr: any[], index: number): 'left' | 'center' | 'right' => {
    if(index===0) {
        return 'left'
    } else if(index<arr.length-1) {
        return 'center'
    } else {
        return 'right'
    }
}
export const renderHistory = (data: DataHistoryType) => {
    const {
        id,
        title,
        event_date_utc,
        event_date_unix,
        details,
        links: { reddit, article, wikipedia }
    } = data as DataHistoryType
    const linksArr = [
        {
            link: reddit,
            label: 'Reddit'
        },
        {
            link: article,
            label: 'Article'
        },
        {
            link: wikipedia,
            label: 'Wikipedia'
        }
    ].filter(lo => lo.link !== null)
    return (<div>
        <DataWrapper>
            <Title>{title}</Title>
            <EventDateUnix>{event_date_unix}</EventDateUnix>
        </DataWrapper>
        <DataWrapper>
            <EventDateUTC>{event_date_utc}</EventDateUTC>
        </DataWrapper>
        <DataWrapper>
            <Details>{details}</Details>
        </DataWrapper>
        <DataWrapper>
            <Links>
                {linksArr.map((linkObj, index) => 
                    <LinkWithTaget 
                        style={{ textAlign: getAlign(linksArr, index) }} 
                        href={linkObj.link}
                    >
                        {linkObj.label}
                    </LinkWithTaget>
                )}
            </Links>
        </DataWrapper>
    </div>
    )
}

export const renderLaunches = (data: DataLaunchType) => {
    const {
        flight_number,
        mission_name,
        launch_date_utc,
        launch_site
    } = data as DataLaunchType

    return (
    <div>
        <DataWrapper>
            <Title>{flight_number}</Title>
            <EventDateUnix>{launch_site && launch_site.site_name}</EventDateUnix>
        </DataWrapper>
        <DataWrapper>
            <EventDateUTC>{launch_date_utc}</EventDateUTC>
        </DataWrapper>
        <DataWrapper>
            <Details>{mission_name}</Details>
        </DataWrapper>
    </div>
    )
}
