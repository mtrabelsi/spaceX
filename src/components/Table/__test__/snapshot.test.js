import React from 'react';
import renderer from 'react-test-renderer';
import Table from '..';

it('Should consider all props and generate unique html based on that', () => {
  const tree = renderer
    .create(
      <Table
        abstractData={[{ id: '1' }, { id: '2' }]}
        itemClickHandler={(id) => console.log(`clicked ${id}`)}
        dataType="history"
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
