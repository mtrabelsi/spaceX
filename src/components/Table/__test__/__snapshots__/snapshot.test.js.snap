// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should consider all props and generate unique html based on that 1`] = `<section />`;
